packages/network-controller/src/NetworkController.ts

```diff
@@ -641,6 +641,19 @@ export class NetworkController extends BaseControllerV2<
     );
   }
 
+  /**
+   * Returns the network client for a given network client id.
+   *
+   * @param networkClientId - The network client id for which to retrieve the network client.
+   * @returns The network client.
+   */
+  getNetworkClientById(
+    networkClientId: NetworkClientId,
+  ): AutoManagedNetworkClient<NetworkClientConfiguration> {
+    const networkClients = this.getNetworkClients();
+    return networkClients[networkClientId];
+  }
```

Comment:
Since we currently have a split between Infura and custom network clients, I think it would be a good idea for this method to return the right type depending on the ID given so that `networkClient.configuration` will return a correctly typed configuration object. Instead of relying on `getNetworkClients`, we can access the registry directly. Also, considering that CustomNetworkClientId is a string, this method can technically return undefined if that string isn't present as a property in the custom network client registry. So consumers don't have to deal with undefined, I think it would be a good idea to throw in that case.

With that in mind, what about this?

``` suggestion
  /**
   * Returns the Infura network client with the given ID.
   *
   * @param infuraNetworkClientId - An Infura network client ID.
   * @returns The Infura network client.
   * @throws If an Infura network client does not exist with the given ID.
   */
  getNetworkClientById(
    infuraNetworkClientId: BuiltInNetworkClientId,
  ): AutoManagedNetworkClient<InfuraNetworkClientConfiguration>;

  /**
   * Returns the custom network client with the given ID.
   *
   * @param customNetworkClientId - A custom network client ID.
   * @returns The custom network client.
   * @throws If a custom network client does not exist with the given ID.
   */
  getNetworkClientById(
    customNetworkClientId: CustomNetworkClientId,
  ): AutoManagedNetworkClient<CustomNetworkClientConfiguration>;

  getNetworkClientById(
    networkClientId: NetworkClientId,
  ): AutoManagedNetworkClient<NetworkClientConfiguration> {
    const autoManagedNetworkClientRegistry =
      this.#ensureAutoManagedNetworkClientRegistryPopulated();

    if (isInfuraProviderType(networkClientId)) {
      const infuraNetworkClient =
        autoManagedNetworkClientRegistry[NetworkClientType.Infura][
          networkClientId
        ];
      if (!infuraNetworkClient) {
        throw new Error(
          `No Infura network client was found with the ID "${networkClientId}".`,
        );
      }
      return infuraNetworkClient;
    }

    const customNetworkClient =
      autoManagedNetworkClientRegistry[NetworkClientType.Custom][
        networkClientId
      ];
    if (!customNetworkClient) {
      throw new Error(
        `No custom network client was found with the ID "${networkClientId}".`,
      );
    }
    return customNetworkClient;
  }
```

With these changes, this should be the case:

``` javascript
const goerliNetworkClient = networkController.getNetworkClientById(InfuraNetworkType.goerli);
goerliNetworkClient.configuration  //=> InfuraNetworkClientConfiguration
// Assuming that 'AAAA-AAAA-AAAA-AAAA' exists in the registry
const customNetworkClient = networkController.getNetworkClientById('AAAA-AAAA-AAAA-AAAA');
customNetworkClient.configuration  //=> CustomNetworkClientConfiguration
// Assuming that 'BBBB-BBBB-BBBB-BBBB' does not exist in the registry
networkController.getNetworkClientById('BBBB-BBBB-BBBB-BBBB');  // throws

Reply:
> if (!infuraNetworkClient) {
        throw new Error(
          `No Infura network client was found with the ID "${networkClientId}".`,
        );
      }
      
I'm pretty sure this couldn't ever be hit the way the Controller is written today since if it entered the if block
>  if (isInfuraProviderType(networkClientId))

it must be a valid infura clientId and must therefore be included in the registry.

I suppose this could be refactored at some point such that this wouldn't be the case so I don't mind leaving the error in.

Reply:
cc @mcmire

Reply:
That's true. I've only added that check because eventually [we want to enable TypeScript's `exactOptionalPropertyTypes` option](https://github.com/MetaMask/core/pull/796), and once that happen it'll add `| undefined` to all index access, so it's good to keep that in mind now.